src/inverse/SIA2D/adjoint.jl

```diff
@@ -0,0 +1,415 @@
+    ∇Sy = Huginn.avg_x(dSdy)
+
+    # Compute surface slope
+    ∇S = (∇Sx.^2 .+ ∇Sy.^2).^(1/2)
+
+    # Compute average ice thickness
+    H̄ = Huginn.avg(H)
+
+    # Compute diffusivity based on target objective
+    D = target.D(
+        H = H̄, ∇S = ∇S, θ = θ,
+        ice_model = SIA2D_model, ml_model = ml_model,
+        glacier = glacier, params = params
+    )
+
+    # TODO: Clip dSdx and dSdx for conservation of mass condition
+
+    ### Computation of partial derivatives of diffusivity
+    ∂D∂H_dual = target.∂D∂H(
+        H = H̄, ∇S = ∇S, θ = θ,
+        ice_model = SIA2D_model, ml_model = ml_model,
+        glacier = glacier, params = params
+    )
+    ∂D∂H = Huginn.avg(∂D∂H_dual)
+
+    β = target.∂D∂∇H(
+        H = H̄, ∇S = ∇S, θ = θ,
+        ice_model = SIA2D_model, ml_model = ml_model,
+        glacier = glacier, params = params
+    )
+
+    # Derivatives evaluated in dual grid
+    ∂D∂∇H_x = β .* ∇Sx
+    ∂D∂∇H_y = β .* ∇Sy
+
+    ### Computation of term ∇⋅(D⋅∇λ)
+
+    # Compute flux components
+    @views dλdx_edges = Huginn.diff_x(λ[:,2:end - 1]) ./ Δx
+    @views dλdy_edges = Huginn.diff_y(λ[2:end - 1,:]) ./ Δy
+
+    Fx = .-Huginn.avg_y(D) .* dλdx_edges
+    Fy = .-Huginn.avg_x(D) .* dλdy_edges
+
+    Fxx = Huginn.diff_x(Fx) / Δx
+    Fyy = Huginn.diff_y(Fy) / Δy
+
+    ### Flux divergence
+    ∇D∇λ = .-(Fxx .+ Fyy)
+
+    ### Computation of term ∂D∂H x ⟨∇S, ∇λ⟩
+    ∇λ∇S_x_edges = dSdx .* Huginn.diff_x(λ) ./ Δx
+    ∇λ∇S_y_edges = dSdy .* Huginn.diff_y(λ) ./ Δy
+    ∇λ∇S_x = Huginn.avg_y(∇λ∇S_x_edges)
+    ∇λ∇S_y = Huginn.avg_x(∇λ∇S_y_edges)
+    ∇λ∇S = ∇λ∇S_x .+ ∇λ∇S_y
+
+    ∂D∂H_∇S_∇λ = ∂D∂H .* Huginn.avg(∇λ∇S)
+
+    ### Computation of term ∇⋅(∂D∂∇H x ⟨∇S, ∇λ⟩)
+    ∂D∂∇H_∇S_∇λ_x = ∇λ∇S .* ∂D∂∇H_x
+    ∂D∂∇H_∇S_∇λ_y = ∇λ∇S .* ∂D∂∇H_y
+
+    ∇_∂D∂∇H_∇S_∇λ_x = Huginn.diff_x(∂D∂∇H_∇S_∇λ_x) ./ Δx
+    ∇_∂D∂∇H_∇S_∇λ_y = Huginn.diff_y(∂D∂∇H_∇S_∇λ_y) ./ Δy
+    ∇_∂D∂∇H_∇S_∇λ = Huginn.avg_y(∇_∂D∂∇H_∇S_∇λ_x) .+ Huginn.avg_x(∇_∂D∂∇H_∇S_∇λ_y)
+
+    ### Final computation of VJP
+    dλ = zero(λ)
+    Huginn.inn(dλ) .= .+ ∇D∇λ .- ∂D∂H_∇S_∇λ .+ ∇_∂D∂∇H_∇S_∇λ
+    return dλ
+end
+
+"""
+    VJP_λ_∂SIA∂θ_continuous(
+        θ,
+        λ::Matrix{R},
+        H::Matrix{R},
+        simulation::SIM,
+        t::R;
+        batch_id::Union{Nothing, I} = nothing
+    )
+Implementation of the continuous adjoint of the SIA2D equation with respect to θ.
+Given λ, H and θ, it returns the VJP of λ^T * ∂(SIA2D)/∂θ (θ).
+
+Arguments:
+- `θ`: Vector of parameters
+- `λ::Matrix{R}`: Adjoint state, also called output gradient in reverse-mode AD.
+- `H::Matrix{R}`: Ice thickness which corresponds to the input state of the SIA2D.
+- `simulation::SIM`: Simulation parameters.
+- `t::R`: Time value, not used as SIA2D is time independent.
+- `batch_id::Union{Nothing, I}`: Batch index.
+
+Returns:
+- `dλ::Matrix{R}`: Jacobian vector product, also called input gradient in reverse-mode AD.
+"""
+function VJP_λ_∂SIA∂θ_continuous(
```

Comment:
Just out of curiosity, but do these need to be out-of-place or they could potentially work in-place?